File: src/admin/router.ts
/**
 * Admin REST API router — exposes role management endpoints.
 *
 * Each service mounts this router to expose standard authorization
 * management endpoints under /api/{service}/v1/authz/.
 */

import type { Router, Request, Response } from 'express';
import { RoleRepository } from '../roles/repository';
import { RoleAssignmentManager } from '../roles/assignment';
import { PermissionResolver } from '../roles/resolver';
import { PermissionChecker } from '../permissions/checker';
import { MembershipCache } from '../membership/cache';
import { GatedContext, RoleDefinition } from '../types';
import { createLogger } from '../logger';

const logger = createLogger('admin-router');

export interface AdminRouterDeps {
  roleRepo: RoleRepository;
  roleAssignments: RoleAssignmentManager;
  permissionResolver: PermissionResolver;
  permissionChecker: PermissionChecker;
  membershipCache: MembershipCache;
}

/**
 * Create the admin router with authorization management endpoints.
 * Requires Express Router to be passed in (no direct express dependency).
 */
export function createAdminRouter(
  createRouter: () => Router,
  deps: AdminRouterDeps,
  service: string,
): Router {
  const router = createRouter();
  const authzPermission = `${service}:authz:manage`;

  // Helper to enforce authz management permission
  function requireAuthzManage(req: Request, res: Response): GatedContext | null {
    const ctx = req.gatedContext;
    if (!ctx) {
      res.status(401).json({ error: 'Authentication required' });
      return null;
    }
    const result = deps.permissionChecker.check(ctx, authzPermission);
    if (!result.allowed) {
      res.status(403).json({ error: 'Forbidden' });
      return null;
    }
    return ctx;
  }

  // ─── Role Endpoints ──────────────────────────────────────────

  // List available roles
  router.get('/roles', async (req: Request, res: Response) => {
    const ctx = requireAuthzManage(req, res);
    if (!ctx) return;

    try {
      const roles = await deps.roleRepo.listForOrg(ctx.org.id);
      res.json({ data: roles });
    } catch (err) {
      logger.error({ err }, 'Failed to list roles');
      res.status(500).json({ error: 'Internal error' });
    }
  });

  // Get role details
  router.get('/roles/:roleId', async (req: Request, res: Response) => {
    const ctx = requireAuthzManage(req, res);
    if (!ctx) return;

    try {
      const role = await deps.roleRepo.resolve(ctx.org.id, req.params.roleId);
      if (!role) {
        res.status(404).json({ error: 'Role not found' });
        return;
      }
      res.json({ data: role });
    } catch (err) {
      logger.error({ err }, 'Failed to get role');
      res.status(500).json({ error: 'Internal error' });
    }
  });

  // Create custom role
  router.post('/roles', async (req: Request, res: Response) => {
    const ctx = requireAuthzManage(req, res);
    if (!ctx) return;

    try {
      const { key, name, description, permissions, inherits } = req.body as RoleDefinition;
      if (!key || !name) {
        res.status(400).json({ error: 'key and name are required' });
        return;
      }

      const role = await deps.roleRepo.create(ctx.org.id, {
        key,
        name,
        description,
        permissions: permissions ?? [],
        inherits: inherits ?? [],
      });
      res.status(201).json({ data: role });
    } catch (err) {
      logger.error({ err }, 'Failed to create role');
      res.status(500).json({ error: 'Internal error' });
    }
  });

  // Update custom role
  router.patch('/roles/:roleId', async (req: Request, res: Response) => {
    const ctx = requireAuthzManage(req, res);
    if (!ctx) return;

    try {
      const role = await deps.roleRepo.update(
        ctx.org.id,
        req.params.roleId,
        req.body,
      );
      if (!role) {
        res.status(404).json({ error: 'Role not found or is a system role' });
        return;
      }
      res.json({ data: role });
    } catch (err) {
      logger.error({ err }, 'Failed to update role');
      res.status(500).json({ error: 'Internal error' });
    }
  });

  // Delete custom role
  router.delete('/roles/:roleId', async (req: Request, res: Response) => {
    const ctx = requireAuthzManage(req, res);
    if (!ctx) return;

    try {
      const deleted = await deps.roleRepo.delete(ctx.org.id, req.params.roleId);
      if (!deleted) {
        res.status(404).json({ error: 'Role not found or is a system role' });
        return;
      }
      res.status(204).send();
    } catch (err) {
      logger.error({ err }, 'Failed to delete role');
      res.status(500).json({ error: 'Internal error' });
    }
  });

  // ─── Member Role Endpoints ───────────────────────────────────

  // List roles for a member
  router.get(
    '/members/:membershipId/roles',
    async (req: Request, res: Response) => {
      const ctx = requireAuthzManage(req, res);
      if (!ctx) return;

      try {
        const assignments = await deps.roleAssignments.forMembership(
          req.params.membershipId,
        );
        res.json({ data: assignments });
      } catch (err) {
        logger.error({ err }, 'Failed to list member roles');
        res.status(500).json({ error: 'Internal error' });
      }
    },
  );

  // Assign a role to a member
  router.post(
    '/members/:membershipId/roles',
    async (req: Request, res: Response) => {
      const ctx = requireAuthzManage(req, res);
      if (!ctx) return;

      try {
        const { role_id } = req.body as { role_id: string };
        if (!role_id) {
          res.status(400).json({ error: 'role_id is required' });
          return;
        }

        // Verify role exists
        const role = await deps.roleRepo.resolve(ctx.org.id, role_id);
        if (!role) {
          res.status(404).json({ error: 'Role not found' });
          return;
        }

        await deps.roleAssignments.assign(
          req.params.membershipId,
          role_id,
          ctx.org.id,
          ctx.identity.id,
        );

        // Rebuild permissions
        const cached = await deps.membershipCache.findById(
          req.params.membershipId,
        );
        if (cached) {
          await deps.permissionResolver.rebuildForMembership(
            req.params.membershipId,
            ctx.org.id,
            cached.groups,
          );
        }

        res.status(201).json({ data: { membership_id: req.params.membershipId, role_id } });
      } catch (err) {
        logger.error({ err }, 'Failed to assign role');
        res.status(500).json({ error: 'Internal error' });
      }
    },
  );

  // Revoke a role from a member
  router.delete(
    '/members/:membershipId/roles/:roleId',
    async (req: Request, res: Response) => {
      const ctx = requireAuthzManage(req, res);
      if (!ctx) return;

      try {
        const revoked = await deps.roleAssignments.revoke(
          req.params.membershipId,
          req.params.roleId,
        );
        if (!revoked) {
          res.status(404).json({ error: 'Role assignment not found' });
          return;
        }

        // Rebuild permissions
        const cached = await deps.membershipCache.findById(
          req.params.membershipId,
        );
        if (cached) {
          await deps.permissionResolver.rebuildForMembership(
            req.params.membershipId,
            ctx.org.id,
            cached.groups,
          );
        }

        res.status(204).send();
      } catch (err) {
        logger.error({ err }, 'Failed to revoke role');
        res.status(500).json({ error: 'Internal error' });
      }
    },
  );

  // ─── Group Role Endpoints ────────────────────────────────────

  // List roles for a group
  router.get(
    '/groups/:groupId/roles',
    async (req: Request, res: Response) => {
      const ctx = requireAuthzManage(req, res);
      if (!ctx) return;

      try {
        const assignments = await deps.roleAssignments.forGroup(
          req.params.groupId,
        );
        res.json({ data: assignments });
      } catch (err) {
        logger.error({ err }, 'Failed to list group roles');
        res.status(500).json({ error: 'Internal error' });
      }
    },
  );

  // Assign a role to a group
  router.post(
    '/groups/:groupId/roles',
    async (req: Request, res: Response) => {
      const ctx = requireAuthzManage(req, res);
      if (!ctx) return;

      try {
        const { role_id } = req.body as { role_id: string };
        if (!role_id) {
          res.status(400).json({ error: 'role_id is required' });
          return;
        }

        await deps.roleAssignments.assignToGroup(
          req.params.groupId,
          role_id,
          ctx.org.id,
          ctx.identity.id,
        );

        res.status(201).json({ data: { group_id: req.params.groupId, role_id } });
      } catch (err) {
        logger.error({ err }, 'Failed to assign role to group');
        res.status(500).json({ error: 'Internal error' });
      }
    },
  );

  // Revoke a role from a group
  router.delete(
    '/groups/:groupId/roles/:roleId',
    async (req: Request, res: Response) => {
      const ctx = requireAuthzManage(req, res);
      if (!ctx) return;

      try {
        const revoked = await deps.roleAssignments.revokeFromGroup(
          req.params.groupId,
          req.params.roleId,
        );
        if (!revoked) {
          res.status(404).json({ error: 'Group role assignment not found' });
          return;
        }

        res.status(204).send();
      } catch (err) {
        logger.error({ err }, 'Failed to revoke role from group');
        res.status(500).json({ error: 'Internal error' });
      }
    },
  );

  // ─── Permission Check Endpoint ───────────────────────────────

  // Check permissions (for admin UIs)
  router.post('/check', async (req: Request, res: Response) => {
    const ctx = requireAuthzManage(req, res);
    if (!ctx) return;

    try {
      const { membership_id, permissions } = req.body as {
        membership_id: string;
        permissions: string[];
      };

      if (!membership_id || !permissions?.length) {
        res.status(400).json({
          error: 'membership_id and permissions are required',
        });
        return;
      }

      // Resolve the target membership's permissions
      const cached = await deps.membershipCache.findById(membership_id);
      if (!cached) {
        res.status(404).json({ error: 'Membership not found' });
        return;
      }

      const resolvedPerms = await deps.permissionResolver.resolvePermissions(
        membership_id,
        ctx.org.id,
        cached.groups,
      );

      const results: Record<string, { allowed: boolean; source: string | null }> = {};
      for (const perm of permissions) {
        const check = deps.permissionChecker.check(
          { ...ctx, membership: { ...ctx.membership, id: membership_id }, permissions: resolvedPerms },
          perm,
        );
        results[perm] = check;
      }

      res.json({
        data: {
          [membership_id]: results,
        },
      });
    } catch (err) {
      logger.error({ err }, 'Failed to check permissions');
      res.status(500).json({ error: 'Internal error' });
    }
  });

  return router;
}

File: sdk-python/gatedhouse/admin/router.py
"""Admin REST API router for role and permission management (FastAPI)."""

from __future__ import annotations

import logging
from typing import Any, TYPE_CHECKING

logger = logging.getLogger("gatedhouse.admin")


def create_admin_router(gatedhouse: Any) -> Any:
    """Create a FastAPI APIRouter with admin endpoints for role management.

    Args:
        gatedhouse: The Gatedhouse orchestrator instance.

    Returns:
        A FastAPI APIRouter instance.
    """
    from fastapi import APIRouter, HTTPException
    from pydantic import BaseModel

    router = APIRouter(prefix="/gatedhouse/admin", tags=["gatedhouse-admin"])

    class RoleCreateRequest(BaseModel):
        key: str
        name: str
        description: str | None = None
        permissions: list[str] = []
        inherits: list[str] = []

    class RoleUpdateRequest(BaseModel):
        name: str
        description: str | None = None
        permissions: list[str] = []
        inherits: list[str] = []

    class RoleAssignRequest(BaseModel):
        membership_id: str
        role_id: str
        assigned_by: str | None = None

    class GroupRoleAssignRequest(BaseModel):
        group_id: str
        role_id: str
        assigned_by: str | None = None

    @router.get("/roles/{org_id}")
    async def list_roles(org_id: str) -> list[dict[str, Any]]:
        roles = await gatedhouse.role_repo.list_for_org(org_id)
        return [
            {
                "id": r.id,
                "name": r.name,
                "description": r.description,
                "permissions": r.permissions,
                "inherits": r.inherits,
                "is_system": r.is_system,
            }
            for r in roles
        ]

    @router.post("/roles/{org_id}")
    async def create_role(org_id: str, body: RoleCreateRequest) -> dict[str, Any]:
        from gatedhouse.core.types import RoleDefinition
        role = RoleDefinition(
            key=body.key, name=body.name, description=body.description,
            permissions=body.permissions, inherits=body.inherits,
        )
        stored = await gatedhouse.role_repo.create(org_id, role)
        return {"id": stored.id, "name": stored.name}

    @router.get("/roles/{org_id}/{role_id}")
    async def get_role(org_id: str, role_id: str) -> dict[str, Any]:
        role = await gatedhouse.role_repo.find_by_id(org_id, role_id)
        if not role:
            raise HTTPException(status_code=404, detail="Role not found")
        return {
            "id": role.id,
            "name": role.name,
            "description": role.description,
            "permissions": role.permissions,
            "inherits": role.inherits,
            "is_system": role.is_system,
        }

    @router.put("/roles/{org_id}/{role_id}")
    async def update_role(
        org_id: str, role_id: str, body: RoleUpdateRequest
    ) -> dict[str, Any]:
        from gatedhouse.core.types import RoleDefinition
        role = RoleDefinition(
            key=role_id, name=body.name, description=body.description,
            permissions=body.permissions, inherits=body.inherits,
        )
        stored = await gatedhouse.role_repo.update(org_id, role_id, role)
        if not stored:
            raise HTTPException(status_code=404, detail="Role not found")
        return {"id": stored.id, "name": stored.name}

    @router.delete("/roles/{org_id}/{role_id}")
    async def delete_role(org_id: str, role_id: str) -> dict[str, str]:
        deleted = await gatedhouse.role_repo.delete(org_id, role_id)
        if not deleted:
            raise HTTPException(status_code=404, detail="Role not found or is system role")
        return {"status": "deleted"}

    @router.post("/assignments/{org_id}")
    async def assign_role(org_id: str, body: RoleAssignRequest) -> dict[str, str]:
        await gatedhouse.role_assignments.assign(
            body.membership_id, body.role_id, org_id, body.assigned_by
        )
        return {"status": "assigned"}

    @router.delete("/assignments/{org_id}/{membership_id}/{role_id}")
    async def revoke_role(
        org_id: str, membership_id: str, role_id: str
    ) -> dict[str, str]:
        await gatedhouse.role_assignments.revoke(membership_id, role_id)
        return {"status": "revoked"}

    @router.post("/group-assignments/{org_id}")
    async def assign_group_role(org_id: str, body: GroupRoleAssignRequest) -> dict[str, str]:
        await gatedhouse.role_assignments.assign_to_group(
            body.group_id, body.role_id, org_id, body.assigned_by
        )
        return {"status": "assigned"}

    @router.delete("/group-assignments/{org_id}/{group_id}/{role_id}")
    async def revoke_group_role(
        org_id: str, group_id: str, role_id: str
    ) -> dict[str, str]:
        await gatedhouse.role_assignments.revoke_from_group(group_id, role_id)
        return {"status": "revoked"}

    @router.get("/permissions/{membership_id}")
    async def get_permissions(membership_id: str) -> dict[str, Any]:
        resolved = await gatedhouse.permission_resolver.get_cached_permissions(membership_id)
        return {
            "membership_id": membership_id,
            "permissions": [
                {"permission": r.permission, "source": r.source}
                for r in resolved
            ],
        }

    @router.post("/permissions/{membership_id}/rebuild")
    async def rebuild_permissions(membership_id: str, org_id: str) -> dict[str, Any]:
        cached = await gatedhouse.membership_cache.find_by_id(membership_id)
        groups = cached.groups if cached else []
        permissions = await gatedhouse.permission_resolver.rebuild_for_membership(
            membership_id, org_id, groups
        )
        return {"membership_id": membership_id, "count": len(permissions)}

    return router
